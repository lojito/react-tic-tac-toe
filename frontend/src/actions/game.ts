import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
