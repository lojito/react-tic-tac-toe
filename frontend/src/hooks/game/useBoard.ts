import { useCallback, useEffect, useRef, useState } from "react";
import Board from "../../classes/Board";
import Computer from "../../classes/Computer";
import {
  Categories,
  DBGame,
  FirstPlayer,
  GameResult,
  ImageName,
  Player,
  PlayingLevel,
  SquareLocation,
} from "../../types";

interface Props {
  disabled: boolean;
  first: FirstPlayer;
  level: PlayingLevel;
  userImage: ImageName;
  computerImage: ImageName;
  category: Categories;
}

function useBoard(
  {
    disabled,
    first,
    level,
    userImage,
    computerImage,
    category: { id: categoryId },
  }: Props,
  onGameOver: (game: DBGame) => void
) {
  const board = useRef(Board);
  const computer = useRef(Computer);
  const gameOver = useRef(false);
  const player = useRef(Player.Nobody);
  const result = useRef(GameResult.Started);
  const winners = useRef<SquareLocation[]>([]);
  board.current.first = first;
  board.current.level = level;
  const [gameBoard, setGameBoard] = useState(board.current.getBoard());

  const playGame = useCallback((square?: SquareLocation) => {
    if (player.current === Player.Computer) {
      computer.current.play();
    } else {
      board.current.place(Player.User, square!);
    }
    setGameBoard(board.current.getBoard());

    winners.current = board.current.isAWinner(player.current);
    if (winners.current.length) {
      gameOver.current = true;
      result.current =
        player.current === Player.Computer ? GameResult.Lost : GameResult.Won;
    } else if (board.current.isFull) {
      gameOver.current = true;
      result.current = GameResult.Draw;
    } else {
      player.current =
        player.current === Player.Computer ? Player.User : Player.Computer;
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, []);

  const resetGame = useCallback(() => {
    board.current.reset();
    gameOver.current = false;
    player.current = Player.Nobody;
    result.current = GameResult.Started;
    winners.current = [];
    setGameBoard(board.current.getBoard());
  }, []);

  useEffect(() => {
    if (!disabled && !gameOver.current) {
      resetGame();
    }
  }, [disabled, resetGame]);

  useEffect(() => {
    if (disabled === false) {
      if (first === Player.Computer) {
        player.current = Player.Computer;
      } else {
        player.current = Player.User;
      }
    } else {
      if (player.current === Player.Computer) {
        playGame();
      }
    }
  }, [disabled, first, playGame]);

  useEffect(() => {
    if (gameOver.current) {
      onGameOver({
        board: gameBoard,
        categoryId,
        first,
        userImage: userImage,
        computerImage: computerImage,
        level,
        result: result.current,
        winners: winners.current,
      });
      gameOver.current = false;
    }
  });

  return {
    playGame,
    gameBoard,
  };
}

export default useBoard;
