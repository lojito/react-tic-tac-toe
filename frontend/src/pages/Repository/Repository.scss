.repository {
  margin-top: 20px;

  select {
     margin: 0 0 20px;
  }
}

.repository-images {
  div {
    width: 23.75%;
    display: inline-block;
    margin-right: 1%;

    img {
      max-width: 100%;
    }
  }

  div:nth-child(4n + 1) {
    margin-left: 1%;
  }
}


@media screen and (max-width: 480px) {
  .repository {
    margin-top: 0;
  }
  
  .repository-images {
    div {
      width: 48.5%;
    }

    div:nth-child(2n + 1) {
      margin-left: 1%;
    }
  }
}
