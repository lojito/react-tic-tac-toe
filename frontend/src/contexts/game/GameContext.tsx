import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
