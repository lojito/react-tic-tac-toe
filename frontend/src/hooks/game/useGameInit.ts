import { useCallback } from "react";
import { changeDisabled, changeOver, changeResult } from "../../actions/game";
import { Action } from "../../actions/types";
import { GameResult } from "../../types";

function useGameInit(dispatch: React.Dispatch<Action>) {
  const handleGameInit = useCallback(() => {
    dispatch(changeOver(false));
    dispatch(changeDisabled(false));
    dispatch(changeResult(GameResult.Started));
  }, [dispatch]);

  return { handleGameInit };
}

export default useGameInit;
