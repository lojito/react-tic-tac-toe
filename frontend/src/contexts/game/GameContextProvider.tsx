import React, { FC, useMemo, useReducer } from "react";
import gameReducer from "../../reducers/gameReducer";
import { GameContext } from "./GameContext";
import initialState from "./GameInitialState";

const GameContextProvider: FC = ({ children }) => {
  const [game, dispatch] = useReducer(gameReducer, initialState);
  const value = useMemo(() => {
    return { game, dispatch };
  }, [game]);

  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
};

export default GameContextProvider;
