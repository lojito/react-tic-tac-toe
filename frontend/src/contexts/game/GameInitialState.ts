import { State } from "../../contexts/game/GameContext";
import categories from "../../data/json/categories.json";
import {
  Categories,
  GameResult,
  ImageName,
  Player,
  PlayingLevel,
} from "../../types";

const NUMBER_OF_CATEGORIES = 10;
const category = categories[
  Math.floor(Math.random() * NUMBER_OF_CATEGORIES)
] as Categories;

const IMAGES_PER_CATEGORY = 20;
const userImage = Math.floor(Math.random() * IMAGES_PER_CATEGORY) as ImageName;
const computerImage = Math.floor(
  Math.random() * IMAGES_PER_CATEGORY
) as ImageName;

const initialState: State = {
  category,
  userImage,
  computerImage,
  first: Player.User,
  level: PlayingLevel.Easy,
  disabled: false,
  over: false,
  result: GameResult.Started,
  winners: [],
};

export default initialState;
