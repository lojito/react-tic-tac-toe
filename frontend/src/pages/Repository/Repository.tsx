import React, { FC, useContext } from "react";
import { GameContext } from "../../contexts/game/GameContext";
import Category from "../../components/Category/Category";
import "./Repository.scss";

const Repository: FC = () => {
  const { game } = useContext(GameContext);
  const folder = game.category.folder;
  const url = `${process.env.PUBLIC_URL}/images`;

  return (
    <div className="repository" data-testid="repository">
      <Category />
      <div className="repository-images" data-testid="images">
        {Array(~~20)
          .fill(1)
          .map((value, index) => {
            return (
              <div key={`image-${index}`}>
                <img src={`${url}/${folder}/${index}.jpg`} alt="user" />
              </div>
            );
          })}
      </div>
    </div>
  );
};

export default Repository;
