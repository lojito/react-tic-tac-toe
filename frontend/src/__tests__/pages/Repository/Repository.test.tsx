import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import { GameContext, State } from "../../../contexts/game/GameContext";
import Repository from "../../../pages/Repository/Repository";

jest.mock("../../../components/Category/Category", () => () => <select />);

describe("Repository", () => {
  const renderRepository = () => {
    const game = {
      category: {
        id: 0,
        folder: "habana",
        name: "CATEGORY_HAVANA_LANDMARKS",
      },
    } as State;

    const dispatch = jest.fn();

    const utils = render(
      <GameContext.Provider value={{ game, dispatch }}>
        <Repository />
      </GameContext.Provider>
    );

    const divWrapper = screen.getByTestId("repository");
    const divImages = screen.getByTestId("images");
    const images = screen.getAllByAltText("user");

    return { ...utils, divWrapper, divImages, images };
  };

  it("renders the Repository component correctly", () => {
    const { asFragment } = renderRepository();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should apply the repository class name", () => {
    const { divWrapper } = renderRepository();

    expect(divWrapper).toHaveClass("repository");
  });

  it("should apply the repository-images class name", () => {
    const { divImages } = renderRepository();

    expect(divImages).toHaveClass("repository-images");
  });

  it("should display the right number of images", () => {
    const { images } = renderRepository();

    expect(images).toHaveLength(20);
  });

  it("should apply the right url to the images", () => {
    const url = "/images";

    const { images } = renderRepository();

    for (const image of images) {
      expect(image).toHaveAttribute("src", expect.stringContaining(url));
    }
  });
});
