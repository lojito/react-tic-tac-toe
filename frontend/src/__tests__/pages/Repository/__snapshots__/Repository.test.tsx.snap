// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository renders the Repository component correctly 1`] = `
<DocumentFragment>
  <div
    class="repository"
    data-testid="repository"
  >
    <select />
    <div
      class="repository-images"
      data-testid="images"
    >
      <div>
        <img
          alt="user"
          src="undefined/images/habana/0.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/1.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/2.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/3.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/4.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/5.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/6.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/7.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/8.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/9.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/10.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/11.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/12.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/13.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/14.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/15.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/16.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/17.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/18.jpg"
        />
      </div>
      <div>
        <img
          alt="user"
          src="undefined/images/habana/19.jpg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
